esp/esdllib/esdl_script.cpp

```diff
@@ -0,0 +1,914 @@
+    {
+#if defined(_DEBUG)
+        DBGLOG(">%s> %s(%s, select('%s'))", m_traceName.str(), tagname.str(), m_target.str(), m_select->getXpath());
+#endif
+    }
+
+    virtual ~CEsdlTransformOperationSetValue(){}
+
+    virtual bool process(IEspContext * context, IPropertyTree *request, IXpathContext * xpathContext) override
+    {
+        if (!xpathContext)
+            throw MakeStringException(-1, "Could not process custom transform (xpathcontext == null)");
+        if (!request)
+            return false;
+        if ((!m_xpath_target && m_target.isEmpty()) || !m_select)
+            return false; //only here if "optional" backward compatible support for now (optional syntax errors aren't actually helpful
+        try
+        {
+            StringBuffer value;
+            xpathContext->evaluateAsString(m_select, value);
+            return doSet(xpathContext, request, value);
+        }
+        catch (IException* e)
+        {
+            StringBuffer msg;
+            esdlOperationError("Custom transform: ", e->errorMessage(msg), m_traceName, !ignoreCodingErrors);
+            e->Release();
+        }
+        catch (...)
+        {
+            esdlOperationError("Custom transform: ", "could not process", m_traceName, !ignoreCodingErrors);
+        }
+        return false;
+    }
+
+    const char *getTargetPath(IXpathContext * xpathContext, StringBuffer &s)
+    {
+        if (m_xpath_target)
+        {
+            xpathContext->evaluateAsString(m_xpath_target, s);
+            return s;
+        }
+        return m_target.str();
+    }
+    virtual bool doSet(IXpathContext * xpathContext, IPropertyTree *tree, const char *value)
+    {
+        StringBuffer xpath;
+        const char *target = getTargetPath(xpathContext, xpath);
+        ensurePTree(tree, target);
+        tree->setProp(target, value);
+        return true;
+    }
+};
+
+class CEsdlTransformOperationAppendValue : public CEsdlTransformOperationSetValue
+{
+public:
+    CEsdlTransformOperationAppendValue(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationSetValue(tree, prefix){}
+
+    virtual ~CEsdlTransformOperationAppendValue(){}
+
+    virtual bool doSet(IXpathContext * xpathContext, IPropertyTree *tree, const char *value)
+    {
+        StringBuffer xpath;
+        const char *target = getTargetPath(xpathContext, xpath);
+        ensurePTree(tree, target);
+        tree->appendProp(target, value);
+        return true;
+    }
+};
+
+class CEsdlTransformOperationAddValue : public CEsdlTransformOperationSetValue
+{
+public:
+    CEsdlTransformOperationAddValue(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationSetValue(tree, prefix){}
+
+    virtual ~CEsdlTransformOperationAddValue(){}
+
+    virtual bool doSet(IXpathContext * xpathContext, IPropertyTree *tree, const char *value)
+    {
+        StringBuffer xpath;
+        const char *target = getTargetPath(xpathContext, xpath);
+        if (tree->getCount(target)==0)
```

Comment:
Are the if/else conditions reversed?

Reply:
no, it's basically saying if it doesn't already exist create the whole path for the first one.  addProp will add new items but not ensure the path.

Reply:
Ah ok in that case it's fine.